concurrency/DigitalItemIonTransform.java

```
@@ -0,0 +1,3117 @@
+   }
+   
+   private static IonStruct getFeatureIon(String name, int major_version, int minor_version,
+         String dataSdl) {
+      IonStruct feature = ION_SYSTEM.newEmptyStruct();
+      addString(feature, FEATURE_NAME, name);
+      addInt(feature, FEATURE_MAJOR_VERSION, major_version);
+      addInt(feature, FEATURE_MINOR_VERSION, minor_version);
+      addStringIfNotNull(feature, FEATURE_DATA_SDL, dataSdl);
+      return feature;
+   }
+
+   @VisibleForTesting
+   protected static void addFeatureList(final TypeSafeIonStruct digitalBook, final DigitalItem item) {
+      if (anyNull(digitalBook.get(FEATURES))) {
+         return;
+      }
+
+      List<Feature> features = new ArrayList<>();
+      IonList featuresIon = digitalBook.getList(FEATURES);
+      for (IonValue fIon: featuresIon) {
+         TypeSafeIonStruct fStruct = IonStructs.newTypeSafeStruct(fIon);
+         Feature feature = new Feature();
+         if ("KW".equals(getStringValue(fStruct, FEATURE_NAME))) {
+            continue;
+         }
+         feature.setName(getStringValue(fStruct, FEATURE_NAME));
+         feature.setSdl(getStringValue(fStruct, FEATURE_DATA_SDL));
+
+         IonInt majorVersion = fStruct.getInt(FEATURE_MAJOR_VERSION);
+         if (!anyNull(majorVersion)) {
+            feature.setMajorVersion(majorVersion.intValue());
+         }
+
+         IonInt minorVersion = fStruct.getInt(FEATURE_MINOR_VERSION);
+         if (!anyNull(minorVersion)) {
+            feature.setMinorVersion(minorVersion.intValue());
+         }
+
+         TypeSafeIonStruct featureConfiguration = fStruct.getStruct(FEATURE_CONFIGURATION);
+         if (!anyNull(featureConfiguration)) {
+            StringBuilder configurationStringBuilder = new StringBuilder();
+            try (IonWriter writer = ION_SYSTEM.newTextWriter(configurationStringBuilder)) {
+               featureConfiguration.writeTo(writer);
+            } catch (IOException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)